### PR TITLE
Bug 1437383: Create User.pm PhabBugz class for loading of a user object from phabricator

### PR DESCRIPTION
**Pull request by dklawren**
This is the User.pm module only for now. I have patches that add support for using the new module for loading of phabricator users instead of the utility methods in Util.pm. I felt review would be easier if I made smaller PRs for each of those instead of one large PR. Please review and once landed, I will create PRs for other parts of the code to utilize the new class module.

dkl

### CODE REVIEW

File: extensions/PhabBugz/lib/User.pm [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    $params->{roles}           = $params->{fields}->{roles};
+    $params->{view_policy}     = $params->{fields}->{policy}->{view};
+    $params->{edit_policy}     = $params->{fields}->{policy}->{edit};
+
+    delete $params->{fields};
+
+    if ( my $external_accounts =
+        $params->{attachments}{'external-accounts'}{'external-accounts'} )
+    {
+        foreach my $account (@$external_accounts) {
```
Comment (dylanwh) [inline extensions/PhabBugz/lib/User.pm]:
```perl
my $external_accounts = $params->{attachments}{'external-accounts'}{'external-accounts'} 
my $bug_user = first { $_->{type} ne 'bmo' } @$external_accounts;
$params->{bug_user_id} = $bug_user->{id};
```

File: extensions/PhabBugz/lib/User.pm [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+use 5.10.1;
+use Moo;
+
+use Bugzilla::User;
+
+use Bugzilla::Extension::PhabBugz::Util qw(request);
+
+use Types::Standard -all;
+use Type::Utils;
```
Comment (dylanwh) [inline extensions/PhabBugz/lib/User.pm]:
```perl
use List::Util qw(first);
```

File: extensions/PhabBugz/lib/User.pm [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+
+    return $phab_users;
+}
+
+#################
+#   Accessors   #
+#################
+
```
Comment (dylanwh) [inline extensions/PhabBugz/lib/User.pm]:
```perl
has 'bugzilla_user' => ( is => 'lazy' );
sub _build_bugzilla_user {
    my ($self) = @_;
   return undef if !$self->bugzilla_id;
   return Bugzilla::User->new( { id => $self->bugzilla_id, cache => 1 } );
}
```

File: extensions/PhabBugz/lib/User.pm [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+sub get_phab_bugzilla_ids {
+    my ( $class, $params ) = @_;
+
+    my $memcache = Bugzilla->memcached;
+
+    # Try to find the values in memcache first
+    my @results;
+    my @bugzilla_ids = @{ $params->{ids} };
```
Comment (dylanwh) [inline extensions/PhabBugz/lib/User.pm]: How big do we expect `@bugzilla_ids` to get? splice() is very slow for large arrays.

**Comment by dklawren** [comment]
Suggested changes made. Please review.

**Comment by zalun** [comment]
Seems finr for me.

**Pull request merged**